Cannot build no_std, missing documentation

I've been trying to compile wasmi without std, specifically a static library with the the c-api. I've been having trouble compiling as it seems as --no-default-features doesn't work.

```
RUSTFLAGS=-Cpanic=abort cargo build -p wasmi_c_api -Zbuild-std=core,alloc -Zjson-target-spec  --target ../powerpc64-ps3-elf.json --no-default-features
...
   Compiling wasmi v2.0.0-beta.2 (/home/benjamin/Desktop/wasmPSL1GHT/wasmi/crates/wasmi)
error[E0463]: can't find crate for `std`
  --> crates/c_api/src/lib.rs:65:1
   |
65 | extern crate std;
   | ^^^^^^^^^^^^^^^^^ can't find crate
   |
   = note: the `powerpc64-ps3-elf` target may not support the standard library
   = help: consider building the standard library from source with `cargo build -Zbuild-std`

```

**TL;DR:** AFAIK you must use `-Z build-std=std` (or `-Z build-std=core,alloc,std`) if you're building with `-Z build-std`, I think this is due to the proc-macro crate but I'm not sure if `wasmi` can do anything about it, might be an issue with `cargo`? (_see below for details_)

---

> have you used `--no-default-features` for the build command to disable default feature in Wasmi.

Their report shows that they used `--no-default-features` and they specifically mentioned that prior to the shared snippet too...? 🤨

---

That said it works fine for a standard target:

```bash
docker run --rm -it --workdir /opt/wasmi ghcr.io/rust-cross/cargo-zigbuild bash
git clone --depth 1 https://github.com/wasmi-labs/wasmi .

RUSTFLAGS='-C panic=abort' cargo zigbuild --release --package wasmi_c_api \
  --no-default-features \
  --target x86_64-unknown-linux-gnu.2.34
```

@nbe1233 is using `-Z json-target-spec --target ../powerpc64-ps3-elf.json` which might be contributing to their problem? (instead of a Tier 1 target)

---

**UPDATE:** When attempting to build the package with nightly and using `-Z build-std=core,alloc`, I produced a similar failure:

```console
$ rustup install nightly --component rust-src --profile minimal
$ cargo +nightly build --package wasmi_c_api --release \
  --no-default-features \
  --target x86_64-unknown-linux-gnu \
  -Z build-std=core,alloc

# ...

   Compiling wasmi_c_api_impl v2.0.0-beta.2 (/opt/wasmi/crates/c_api)
   Compiling wasmi v2.0.0-beta.2 (/opt/wasmi/crates/wasmi)
   Compiling wasmi_c_api_macros v2.0.0-beta.2 (/opt/wasmi/crates/c_api/macro)
error[E0152]: duplicate lang item in crate `core` (which `std` depends on): `sized`
  |
  = note: the lang item is first defined in crate `core` (which `wasmi_c_api` depends on)
  = note: first definition in `core` loaded from /opt/wasmi/target/x86_64-unknown-linux-gnu/release/deps/libcore-28a264df45790277.rmeta
  = note: second definition in `core` loaded from /usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-58aeee91aeb71c13.rmeta
```

This is **avoided if adding `std` into the `-Z build-std` arg** (`-Z build-std=core,std,alloc`).

I'm not sure why that's required exactly, but perhaps the `wasmi_c_api_macros` crate where the build fails is affected?
- It pulls in `std` (pre-compiled) that depends upon `core` but seems to conflict with the two copies? (_pre-compiled vs `-Z build-std` built `core`_).
- Adding `std` to `-Z build-std` somehow avoids that and presumably has no relevance to the target build where `std` wouldn't be linked in?

---

**FWIW:** I can reproduce the same failure @nbe1233 shared, but for a Tier 1 target.

I assume the failure scenario is the same despite different error messages? This one is from targeting a foreign arch (_`aarch64` target vs `x86_64` host, which the proc macro will rely upon_):

Fails with `-Z build-std` slightly differently for a foreign target:

```console
$ cargo +nightly zigbuild --package wasmi_c_api --release \
  --no-default-features \
  --target aarch64-unknown-linux-gnu \
  -Z build-std=core,alloc

# ...

   Compiling wasmi_c_api_impl v2.0.0-beta.2 (/opt/wasmi/crates/c_api)
   Compiling wasmi v2.0.0-beta.2 (/opt/wasmi/crates/wasmi)
   Compiling wasmi_c_api_macros v2.0.0-beta.2 (/opt/wasmi/crates/c_api/macro)
error[E0463]: can't find crate for `std`
  --> crates/c_api/src/lib.rs:65:1
   |
65 | extern crate std;
   | ^^^^^^^^^^^^^^^^^ can't find crate
   |
   = note: the `aarch64-unknown-linux-gnu` target may not support the standard library
   = help: consider building the standard library from source with `cargo build -Zbuild-std`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `wasmi_c_api_impl` (lib) due to 1 previous error
```

Adding `std` will likewise fix that.

And just for added context, without `-Z build-std` it does compile successfully (_not an option for custom targets though AFAIK?_):

```console
# `zigbuild` instead of `build` here to link the foreign arch correctly:
$ cargo +nightly zigbuild --package wasmi_c_api --release \
  --no-default-features \
  --target aarch64-unknown-linux-gnu

# ...

   Compiling wasmi_c_api_macros v2.0.0-beta.2 (/opt/wasmi/crates/c_api/macro)
   Compiling wasmi_c_api v2.0.0-beta.2 (/opt/wasmi/crates/c_api/artifact)
    Finished `release` profile [optimized] target(s) in 1m 27s
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -36,7 +36,7 @@ wasmi_core = { version = "2.0.0-beta.3", path = "crates/core", default-features 
 wasmi_ir = { version = "2.0.0-beta.3", path = "crates/ir", default-features = false }
 wasmi_collections = { version = "2.0.0-beta.3", path = "crates/collections", default-features = false }
 wasmi_wast = { version = "2.0.0-beta.3", path = "crates/wast", default-features = false }
-wasmi_c_api_impl = { version = "2.0.0-beta.3", path = "crates/c_api" }
+wasmi_c_api_impl = { version = "2.0.0-beta.3", path = "crates/c_api", default-features = false }
 wasmi_c_api_macros = { version = "2.0.0-beta.3", path = "crates/c_api/macro" }
 wasmi_fuzz = { version = "2.0.0-beta.3", path = "crates/fuzz" }
 

--- a/crates/c_api/README.md
+++ b/crates/c_api/README.md
@@ -51,3 +51,18 @@ fn main() {
         .compile("your_library");
 }
 ```
+
+## `no_std` Support
+
+The `wasmi_c_api_impl` crate supports `no_std` by disabling its default features.
+
+For `no_std` builds, users may have to define their own `#[global_allocator]` and `#[panic_handler]`
+for the resulting final `staticlib` or `cdylib`.
+
+For targets without a pre-built `core`/`alloc`, build the standard library from
+source with nightly `-Z build-std`. Include `std` in the list (the host
+proc-macro dependency needs it) even though the artifact is `no_std`:
+
+```shell
+cargo +nightly build --no-default-features -Z build-std=core,alloc,std --target <target>
+```

